barman/backup_executor.py

```diff
@@ -658,6 +659,33 @@ def _retry_handler(self, dest_dirs, command, args, kwargs, attempt, exc):
         # Remove all the destination directories and reinit the backup
         self._prepare_backup_destination(dest_dirs)
 
+    def _make_err_handler(self):
+        """
+        Handler invoked during a backup when anything is sent to stderr.
+
+        Used to perform a WAL switch on a primary server if pg_basebackup
+        is running against a standby, otherwise just logs output at INFO
+        level.
+        """
+
+        def err_handler(line):
```

Comment:
I am not a huge fan of functions defined within functions. 
I am also wondering if the self within `err_handler` refers to `PostgresBackupExecutor` instance (in wich case the function is like a disguised object).
But I can see this is heavily used already. I think a more appropriate solution would be to use an interface with handle_error function. and passing anything that implements this interface the the handler would be using this function.
It might not be worth refactoring this behavior, and maybe this solution has limitations. We can probably live with that until we can't :)

Reply:
On second thoughts, I don't think `partial` is going to let me assign the implicit `self`.